### DISCUSSION TITLE
Welcome!

### DISCUSSION THREAD

**Discussion by matheo** [myndpm/open-source#1] [General]
## 👋 Welcome!
Together we can make the Angular Community better 💪
  * Ask questions you’re wondering about.
  * Share ideas and your experience to improve the libraries or build recipes.
  * Engage with other community members.
  * Express your gratitude for having these libraries for free :partying_face:

**Comment by adam-dorin** [comment]
Hey @matheo!
Nice to see that someone else is looking at making some dynamic forms!
My feedback (based on my own approach) for how to handle it would be:

*  Typed configuration: the form receives some sort of configuration from various sources ( API call, config file, directly in the controller)
* Flexibility: users should be able to define their own components for inputs, selects, etc ( the user would be responsible for the update hooks of the dynamic form in this case) my idea was to use a directive and a factory to instantiate the components
* Validations: didn't quite get to implement this yet, but the general idea is to support the normal Validators class in the @angular/forms

**Reply by matheo** [reply]
Nice to see you here!

* Yeah, types are a big plus, and currently an overload Factory ([createMatConfig](https://github.com/myndpm/open-source/blob/master/libs/forms/material/src/dyn-forms-material.factory.ts#L50)) looked like a good way to go for modules to provide an utility, without having to import many component or classes in the App components; it will complain if you use an object not corresponding to the control that you are configuring. Also the abstract [DynControl](https://github.com/myndpm/open-source/blob/master/libs/forms/core/src/dyn-control.class.ts) class is made with 4 generic types to have a strict type check :)
* Indeed. The DynControls have most of the responsibility, the library sets their `control` inside the Form Hierarchy, provide the configurated options (validators) and parameters, and the Dynamic Component should roll by itself from there. With the `TParams` generic I guess is enough for any control to receive the required parameters, perhaps for File Uploads we would need to configure a general token with endpoint/token/etc to impact all the controls, instead providing such configuration in a 1:1 basis (perhaps supporting overrides), just an example that comes to my mind after studying some other dynamic-form objects our there.
* Validators are the upcoming feature, and I've seen that many dynamic-form libraries uses an object like `{ name: args } ` registering the `ValidatorFn|ValidatorFactory` by name in a specific token. I guess Ilike this approach, perhaps we need a solid way to handle error messages and i18n (if someone wants to implement a custom way). Let's talk about validators on the already created discussion.

Glad to read your points, and welcome @adam-dorin!

**Comment by Yumitoya8569** [comment]
Hi guys!

Nice to see here going to build a dynamic forms generic layer library!

**Reply by matheo** [reply]
Welcome to these discussions!
Looking forward to talk with you about `ng-conditional-validator` and more Validators stuff!

**Comment by moolsbytheway** [comment]
Hello everyone,
Nice to see some people trying to solve the world problems.

**Reply by matheo** [reply]
Glad to see you here Moulaye!
that's the thesis! can we solve them with a generic library? :slightly_smiling_face: 
https://dev.to/myndpm/a-new-approach-to-have-dynamic-forms-in-angular-5d11

**Comment by begandroide** [comment]
Hello there! Dynamic forms saved me a lot of working hours, i hope i can share soon.... regards

**Reply by matheo** [reply]
Glad to see you here Benjamín
Is there any special use-case that you need to support with Dynamic Forms?
looking forward to know what do you think of the experience while developing with `@myndpm/dyn-forms` :wink:

**Reply by begandroide** [reply]
hi @matheo , yes I have a special use-case supported with Dynamic Forms, in an organization requested that they want build the content of the forms because always changes their content. 

I build a module with `ng-dynamic-forms` customized to my context, the major features that I constantly need are:
- Auto-calculus of one or more fields, taking as parameter the value of other controls (related form controls)
- Has a FormArray always to add more info to form (like notes)
- Add control to display images and texts.
- Provide custom form controls like rich text editor,
- Auto-save forms at timer (temps)

I need say that all works fine but during development I faced some difficulties, like:
- Controls in FormArray can't specify a layout model or cssClass to only one item of formArray, the change is general (prototype model).
- Model of dynamicFormControl (base) can't be extended because core services doesn't support Generic class overriding.
- Need has related form control with relation "visibility" (to make control invisible, not hidden), principally because at hide, the grid layout moves, and this require more css class logic.

By the other hand, i still don't develop with your organization library, i will try it.

Regards 😄

**Reply by matheo** [reply]
Nice!
I'd like to have a little specification of this use-cases so we can buidl a new demo to add to https://mynd.dev/demos

1. We can add auto-calculus with a custom Matcher, I will think about miltiple sources of changes to facilitate that one :)
2. We have FormArrays, might need to know how do you manage them in the UI
3. Displaying of images and texts seems pretty easy :)
4. A custom control with your preferred rich text editor is also possible
5. This kind of auto-save timer should be implemented outside the form, so your previous solution should still work

About the difficulties you had:

1. dyn-forms has infinite nesting and granular cssClasses per control, so you got covered in this point ;)
2. I need to know which kind of extensiondo you want to implement, by default dyn-forms have strong Generic Types and all the existing controls extends some abstract classes that are DynFormArray, DynFormGroup, DynFormControl and DynFormContainer
3. We got this implemented with the `INVISIBLE` matcher using the CSS `visibility: invisible` :D 

Looks like we can build a cool demo with your experience
Please join us on Discord: https://discord.gg/XxEqkvzeXg
and we keep talking more frequently ;)
